Allow 'in' to terminate expressions in for fixes #5203

rhuanjl opened the pull request:
Attempt at fixing issue #5203 

Based on #5216 

Not 100% sure I've got this right but it passes the mentioned cases and doesn't regress anything I'm aware of.

cc @zenparsing @akroshg

akroshg commented:
@rhuanjl the fix seems correct to me. I'll let @aneeshdk to comment/approve this. Thanks for making the change.

Cellule commented:
I still get a syntax error with
```js
for (var b = () => "" in []) {
  undefined;
}
```

MSLaguana commented on test/es6/generators-syntax.js:
```diff
@@ -212,6 +212,14 @@ var tests = [
+    {
+        name : "yield is allowed before 'in' in a for loop control but not elsewhere bug issue #5203",
+        body : function () {
+            assert.doesNotThrow(function () { eval("function* gf() {for(var a = yield in {});}"); }, "Yield is allowed before 'in' in a for loop");
```
What is the correct interpretation of `for(var a = yield 1 in {}) {…`? I think the only way for that to parse is as `for( var a = (yield 1) in {})`, is that how it behaves with this change?

MSLaguana replied:
And going further afield, it looks like v8 parses ` for(var a = yield 'a' in {} in {a: 1})` as `for(var a = (yield 'a') in ({} in {a:1}))`; does that work out here as well?

rhuanjl replied:
Both of these work in that way - I'll compose some functionality tests.

rhuanjl commented:
I've:
1. Added some functionality tests cases in response to @MSLaguana's points above
2. Updated latest

@Cellule that case follows a different path in the Parser and would require a different fix notably it seems that after parsing the lambda expression ```() => ""``` the parser has not reached the "in" token and is possibly still on the quotes - I'm afraid I'm not sure how/why this happens.

rhuanjl commented:
Please let me know if I should do anything further on this.

akroshg commented on lib/Parser/Parse.h:
```diff
@@ -780,7 +780,7 @@ class Parser
     template<bool buildAST> IdentPtr ParseSuper(bool fAllowCall);
 
-    bool IsTerminateToken();
+    bool IsTerminateToken(bool inTerminate);
```
Nit : would it better to understand if this is called as tkInAsTerminate?

aneeshdk replied:
Can't we just reuse the name fAllowIn?

rhuanjl replied:
I've renamed it to fAllowIn

akroshg commented:
LGTM - @aneeshdk can you do the review and validate?
Should this be rebase to release/1.10?

Cellule commented:
I think it should target release/1.10 yeah

aneeshdk commented:
LGTM. release/1.10 sounds good.

aneeshdk commented on test/es6/forInEdgeCases.js:
```diff
@@ -0,0 +1,80 @@
+                    assert.isTrue(false, "shouldn't reach here");
+                }
+            }
+            testGen(gen3, ['d'], 1);
```
Can we also add a test case where a value is passed into the next call?

rhuanjl replied:
I'm sorry I'm not sure what you're asking for.

The testGen function I'm using a the top uses an assertion to verify that the correct value is yielded - not sure what I should o beyond that?

rhuanjl commented:
I've:
1. updated to target 1.10
2. updated to latest
3. renamed to the inTerminate parameter to fAllowIn per review above

I have not:
1. Added any more tests - @aneeshdk unsure what additional test case you were asking for
2. Fixed the Lambda returning a string case: ``` for (var b = () => "" in []) ``` Possibly should open a separate issue for this - there's something else going on here and I can't work it out - after evaluating the lambda it seems to think that the next token is a double quote, not the in.

Sorry that I was a bit slow to respond to feedback on this.

rhuanjl commented:
This will need updating to be compatible with #5380 as they're touching the same code - there's some overlap both this and #5380 fix:
```js
for (var i = () => {} in {});
```
This also fixes
```js
for (var a = yield 'd' in {} in {a: 1})
```
Which #5380 does not fix.

BUT #5380 fixes:
```js
for (var b = () => "" in [])
```
Which this does not fix.

I assume #5380 will land first then can update this to fit with it.

dilijev commented:
@rhuanjl did you make all the updates you wanted to make after #5380 landed?

rhuanjl commented:
Yes I think this is good to land now.

aneeshdk commented on test/es6/forInEdgeCases.js:
```diff
@@ -0,0 +1,80 @@
+            function* gen1()
+            {
+                for (var a = yield 'a' in {b: 1}) {
+                    assert.isTrue(a == "b");
```
Can you add a flag to make sure the loop is executed?

aneeshdk commented:
LGTM

rhuanjl commented:
Please let me know if there is anything further I need to do on this.

Is it going to be merged? Or should I close it?

dilijev commented:
@rhuanjl Apologies for letting this fall off my radar. You can leave it open. Right now we are re-evaluating which branch to merge this to.

dilijev commented:
@rhuanjl thanks for the contribution!
